lib/gl_rubocop/gl_cops/no_stubbing_perform_async.rb

```diff
@@ -0,0 +1,59 @@
+module GLRubocop
+  module GLCops
+    # This cop ensures that you don't stub perform_async. Stubbing prevents seeing errors from
+    # invalid argumens - because sidekiq arguments have to be valid JSON data types:
+    # github.com/sidekiq/sidekiq/wiki/Best-Practices#1-make-your-job-parameters-small-and-simple
+    #
+    # Good:
+    #   expect(ExampleWorker).to have_enqueued_sidekiq_job
+
+    #
+    # Bad:
+    #   allow(ExampleWorker).to receive(:perform_async)
+    #   expect(ExampleWorker).to receive(:perform_async)
+    #   expect(SomeWorker).not_to have_received(:perform_in)
+    class NoStubbingPerformAsync < RuboCop::Cop::Cop
+      MSG = "Don't stub perform async. Use the rspec-sidekick matchers instead: " \
+            'expect(JobClass).to have_enqueued_sidekiq_job'.freeze
+
+      # This pattern captures expectations for perform_async and perform_in
+      # expect(SomeWorker).not_to have_received(:perform_async)
+      # expect(SomeWorker).to have_received(:perform_async)
+      # expect(SomeWorker).to have_received(:perform_in)
+      def_node_matcher :perform_method_expectation?, <<~PATTERN
+        (send#{' '}
+          (send nil? :expect $_)#{' '}
+          {:not_to :to_not :to}#{' '}
+          (send nil? :have_received (sym {:perform_async :perform_in}) ...))
+      PATTERN
+
+      # This pattern captures allow statements for perform_async and perform_in
+      # allow(SomeWorker).to receive(:perform_async)
+      # allow(SomeWorker).to receive(:perform_in)
+      def_node_matcher :perform_method_allow?, <<~PATTERN
+        (send
+          (send nil? :allow $_)
+          :to
+          (send nil? :receive (sym {:perform_async :perform_in}) ...))
```

Comment:
I see an example of us also stubbing just the perform method [here](https://github.com/givelively/charity-api/blob/b47dfc3d72d2a42f821da920a4a66646bc9e8544/spec/support/capybara.rb#L21). Should that be included as a violation of this rule as well?

Reply:
That seems like a very special case. I can't really imagine why anyone would do that anywhere else - but you're probably correct, that does seem like it is the same sort of violation.

For the record - a number of the places that currently violate this rule, like the place you linked, violate this rule because it's the best way to achieve what they need to achieve, and ought to have this rubocop disabled for their usage.

Reply:
Actually - I take it back, I think stubbing perform is fine, because:

1. The problem with stubbing `perform_async` is that you're tightly coupling the test to the implementation, when you can easily test the actual functionality - almost all the time what you want to test is that a job was added to the queue (not that `perform_async` / `perform_in` was called). In addition, certain things can't be serialized correctly and stubbing the method prevents receiving those errors. Neither of those things is true for `perform` - you aren't adding the job to the queue (so there isn't a better thing to test) and it's directly calling the method so arguments don't get serialized.
2. There are a variety of other classes that have a `perform` method, blocking stubbing/mocking that is unnecessary

Reply:
Yep!